src/vm_cpu.rs

```diff
@@ -95,6 +100,195 @@ pub struct VmCpuTrapState {
     pub htval: u64,
     pub htinst: u64,
 }
+#[derive(Default)]
+#[repr(C)]
+struct VmPmuState {
+    // Stores values of PMU CSRs
+    pmu_csrs: [u64; 32],
+    // Stores an index into PmuCounterInfo for configured counters (max index is 31)
+    configured_counters: u64,
+    started_counters: u64,
+}
+
+impl VmPmuState {
+    fn read_pmu_csr(csr: u64) -> Option<u64> {
+        #[inline]
+        fn read_csr<const V: u16>() -> u64 {
+            let r: u64;
+            unsafe {
```

Comment:
still needs an unsafe comment. Are we _sure_ we can always read CSRs safely? I can't think of a counter example. @abrestic-rivos does that seem OK?

Reply:
> I think we can read all the counter CSRs safely, but there are other CSRs that have read side-effects (e.g. `stopi` from the AIA spec) which we shouldn't go reading blindly.

for sure reading many CSRs can be done incorrectly but will any of them cause memory safety issues?

Reply:
Ah. I can't think of any CSR where a read could have side effects that cause memory safety issues.